### PR TITLE
Auto select server based on Wi-Fi SSID

### LINKED ISSUE #2229: Select server instance based on wifi ssid
Add a setting to the local connection called wifi ssid. When the app is opened, the active server is selected based on the wifi ssid, if configured.

_Originally posted by @mueller-ma in https://github.com/openhab/openhab-android/pull/1900#issuecomment-604945035_

### PR DESCRIPTION
**Pull request by mueller-ma**
An optional Wi-Fi SSID can be configured for each server. If the app is
started and connected to one of these Wi-Fis, it switches to the
corresponding server.

Later this preference could be used force connecting to a remote url,
see #146.

Closes #2229

Signed-off-by: mueller-ma <mueller-ma@users.noreply.github.com>

### CODE REVIEW

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt [hunk 1]
```diff
@@ -524,6 +527,85 @@ class MainActivity : AbstractBaseActivity(), ConnectionFactory.UpdateListener {
...
     }
 
+    private fun selectServerBasedOnWifi() {
+        fun getPermissionForSsidRead() = when {
+            Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q -> Manifest.permission.ACCESS_FINE_LOCATION
+            Build.VERSION.SDK_INT >= Build.VERSION_CODES.O -> Manifest.permission.ACCESS_COARSE_LOCATION
+            else -> null
+        }
+
+        val serverHaveWifiSet = prefs
```
Comment (JohannesPtaszyk) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:537]: I would suggest to prefix this with "does"

File: mobile/src/main/java/org/openhab/habdroid/model/ServerConfiguration.kt [hunk 2]
```diff
@@ -62,7 +62,8 @@ data class ServerConfiguration(
     val localPath: ServerPath?,
     val remotePath: ServerPath?,
     val sslClientCert: String?,
-    val defaultSitemap: DefaultSitemap?
+    val defaultSitemap: DefaultSitemap?,
+    val wifiSsid: String?
```
Comment (JohannesPtaszyk) [inline mobile/src/main/java/org/openhab/habdroid/model/ServerConfiguration.kt:66]: We seem to initialize this with lots of "nulls". I would suggest to insert some default null values.

**Review by JohannesPtaszyk — APPROVED** [review]
Overall this is fine for me. I just added some hints and suggestions where I would make changes :)

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt [hunk 3]
```diff
@@ -524,6 +527,85 @@ class MainActivity : AbstractBaseActivity(), ConnectionFactory.UpdateListener {
         }
     }
 
+    private fun selectServerBasedOnWifi() {
+        fun getPermissionForSsidRead() = when {
```
[the remark this answers is not in the document]
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:531]: It only returns the required permissions for reading the wifi ssid when the app is in foreground. There's already a public function for foreground and background access: `BackgroundTasksManager.getRequiredPermissionsForTask()`.

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt (hunk 1 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:537]: What about `hasAnyServerWifiSet`?

File: mobile/src/main/java/org/openhab/habdroid/model/ServerConfiguration.kt (hunk 2 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/model/ServerConfiguration.kt:66]: I only found two cases where `wifiSsid` is set to `null`: During upgrade from a pre-multi server version and when adding a new server. In latter case every argument except the name is null, so I wouldn't initialize `wifiSsid` with `null` by default.

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt (hunk 1 above)
Comment (maniac103) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt]: `anyServerHasSetWifi`? (Putting `Set` in the middle to not imply this variable actually is a set)

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (JohannesPtaszyk) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:531]: Fine for me :)

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt (hunk 1 above)
Comment (JohannesPtaszyk) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt]: Sounds reasonable

File: mobile/src/main/java/org/openhab/habdroid/model/ServerConfiguration.kt (hunk 2 above)
Comment (JohannesPtaszyk) [reply mobile/src/main/java/org/openhab/habdroid/model/ServerConfiguration.kt:66]: 👍 was not sure if it makes sense. Thanks for clarification!

**Review by JohannesPtaszyk — APPROVED** [review]

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt [hunk 4]
```diff
@@ -242,6 +244,7 @@ class MainActivity : AbstractBaseActivity(), ConnectionFactory.UpdateListener {
         }
 
         EventListenerService.startOrStopService(this)
+        selectServerBasedOnWifi()
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:247]: I'm not sure if `onCreate()` is the best location to call `selectServerBasedOnWifi()`. Maybe do it in `onResume()`, but only if the last time the app was open was some time ago?

Comment (JohannesPtaszyk) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:247]: Why not onCreate tho? 🤔

Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:247]: The user might close the app by pressing the home button, go to a location with a different known wifi, open the app again and it might be still in memory, so `onCreate()` might not be called.

Comment (JohannesPtaszyk) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:247]: Not sure about the permission state on newer Android phones, but wouldn't it make sense to do it in a BroadcastReceiver?

Comment (JohannesPtaszyk) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:247]: Never mind. Thats not possible anymore. Sorry.

Comment (maniac103) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:247]: > Maybe do it in onResume(), but only if the last time the app was open was some time ago?

I'd do it in onStart, and compare current Wifi SSID to the one from the last check ... so base it on actual SSID change, not time.

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt [hunk 5]
```diff
@@ -524,6 +527,85 @@ class MainActivity : AbstractBaseActivity(), ConnectionFactory.UpdateListener {
...
+
+        val anyServerHasSetWifi = prefs
+            .getConfiguredServerIds()
+            .map { id -> ServerConfiguration.load(prefs, getSecretPrefs(), id) }
+            .any { config -> config?.wifiSsid?.isNotEmpty() == true }
+
+        val locationManager = getSystemService(Context.LOCATION_SERVICE) as LocationManager
+        val wifiManager = applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
+        val wifiInfo = wifiManager.connectionInfo
+        val wifiSsid = if (wifiInfo.networkId == -1) null else wifiInfo.ssid.removeSurrounding("\"")
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:545]: Can't you shorten this to `val wifiSsid = wifiManager.connectionInfo.ssid?.removeSurrounding("\"")`?

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt [hunk 6]
```diff
@@ -524,6 +527,85 @@ class MainActivity : AbstractBaseActivity(), ConnectionFactory.UpdateListener {
...
+            wifiSsid.isNullOrEmpty() -> {
+                Log.d(TAG, "Cannot auto select server: SSID empty, probably not connected to wifi")
+                return
+            }
+        }
+
+        val serverForCurrentWifi = prefs
+            .getConfiguredServerIds()
+            .map { id -> ServerConfiguration.load(prefs, getSecretPrefs(), id) }
+            .firstOrNull { config -> config?.wifiSsid == wifiSsid } ?: return
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:585]: Nit: elvis operator on new line?

File: mobile/src/main/java/org/openhab/habdroid/util/ExtensionFuncs.kt [hunk 7]
```diff
@@ -455,13 +456,18 @@ fun ServiceInfo.addToPrefs(context: Context) {
     val port = port.toString()
     Log.d(Util.TAG, "Service resolved: $address port: $port")
 
+    val wifiManager = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
+    val wifiInfo = wifiManager.connectionInfo
+    val wifiSsid = if (wifiInfo.networkId == -1) null else wifiInfo.ssid.removeSurrounding("\"")
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/util/ExtensionFuncs.kt:461]: This seems kind of magical. I think I'd rather leave the SSID empty for automatically created servers.

File: mobile/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -116,6 +116,10 @@
     <string name="settings_server_primary_url" translatable="false">https://www.openhab.org/docs/apps/android.html#multi-server-support</string>
     <string name="settings_server_default_name">openHAB %d</string>
     <string name="click_here_for_more_information">Touch here for more information</string>
+    <string name="settings_multi_server_wifi_ssid_summary">The app will select this server if it\'s connected to this Wi-Fi name</string>
```
Comment (maniac103) [inline mobile/src/main/res/values/strings.xml:119]: Consider on/off strings:
`The app will select this server when connected to Wi-Fi '%s'`
`The app will not automatically select this server based on connected Wi-Fi network`

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt (hunk 4 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:247]: Should the last check be saved in memory or shared prefs?

File: mobile/src/main/java/org/openhab/habdroid/util/ExtensionFuncs.kt (hunk 7 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/util/ExtensionFuncs.kt:461]: If the user manually configures a second server and at some point the app switches to the first server magically, a snackbar is shown to inform the user. Then the user knows about this feature or he has seen it before when adding the second server.

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt (hunk 5 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:545]: `getSSID()` always returns a non-null string:

> The SSID may be {@link WifiManager#UNKNOWN_SSID}, if there is no network currently connected or if the caller has insufficient permissions to access the SSID.

````java
public static final String UNKNOWN_SSID = "<unknown ssid>";
````

File: mobile/src/main/java/org/openhab/habdroid/util/ExtensionFuncs.kt (hunk 7 above)
Comment (maniac103) [reply mobile/src/main/java/org/openhab/habdroid/util/ExtensionFuncs.kt:461]: Still, I don't think second-guessing the user's intent is a good idea.

Comment (maniac103) [reply mobile/src/main/java/org/openhab/habdroid/util/ExtensionFuncs.kt:461]: Maybe a dialog or snackbar asking the user whether he wants to bind the found server to the current WiFi network would be a decent compromise?

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt (hunk 4 above)
Comment (maniac103) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:247]: IMHO memory, since you *do* want this to happen when opening the app when it wasn't running before, regardless of the state before app shutdown.

File: mobile/src/main/res/values/strings.xml (hunk 8 above)
Comment (mueller-ma) [reply mobile/src/main/res/values/strings.xml:119]: Done

File: mobile/src/main/java/org/openhab/habdroid/util/ExtensionFuncs.kt (hunk 7 above)
Comment (JohannesPtaszyk) [reply mobile/src/main/java/org/openhab/habdroid/util/ExtensionFuncs.kt:461]: Guess, since it is important a Dialog would be sufficient 👍

**Comment by maniac103** [comment]
2 commits for your consideration: https://github.com/maniac103/openhab.android/commits/2229-multi-server-wifi

File: mobile/src/main/java/org/openhab/habdroid/util/ExtensionFuncs.kt (hunk 7 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/util/ExtensionFuncs.kt:461]: I'd rather not add the wifi ssid at all instead of showing a dialog. This is probably too special for most users.
